x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java

```diff
@@ -138,6 +152,78 @@ protected Clock getClock() {
+     */
+    public static void checkReadyForXPackCustomMetadata(ClusterState clusterState) {
+        List<DiscoveryNode> notReadyNodes = nodesNotReadyForXPackCustomMetadata(clusterState);
+        if (notReadyNodes.isEmpty() == false) {
+            throw new IllegalStateException("The following nodes are not ready yet for enabling x-pack custom metadata: " + notReadyNodes);
+        }
+    }
+
+    /**
+     * Checks if the cluster state allows this node to add x-pack metadata to the cluster state.
+     * See {@link #checkReadyForXPackCustomMetadata} for more details.
+     */
+    public static boolean isReadyForXPackCustomMetadata(ClusterState clusterState) {
+        return nodesNotReadyForXPackCustomMetadata(clusterState).isEmpty();
+    }
+
+    /**
+     * Returns the list of nodes that won't allow this node from adding x-pack metadata to the cluster state.
+     * See {@link #checkReadyForXPackCustomMetadata} for more details.
+     */
+    public static List<DiscoveryNode> nodesNotReadyForXPackCustomMetadata(ClusterState clusterState) {
+        // check if there's already x-pack metadata in the cluster state; if so, any further metadata won't hurt
+        final MetaData metaData = clusterState.metaData();
+        if (metaData.custom(LicensesMetaData.TYPE) != null ||
+            metaData.custom(MLMetadataField.TYPE) != null ||
+            metaData.custom(WatcherMetaData.TYPE) != null ||
+            clusterState.custom(TokenMetaData.TYPE) != null) {
+            return Collections.emptyList();
+        }
+
+        // if there's no x-pack metadata yet in the cluster state, check that all nodes would be capable
+        // of deserializing newly added x-pack metadata
+        final List<DiscoveryNode> notReadyNodes = StreamSupport.stream(clusterState.nodes().spliterator(), false).filter(node -> {
+            final String xpackInstalledAttr = node.getAttributes().getOrDefault(XPACK_INSTALLED_NODE_ATTR, "false");
+
+            // The node attribute XPACK_INSTALLED_NODE_ATTR was only introduced in 6.3.0, so when
+            // we have an older node in this mixed-version cluster without any x-pack metadata,
+            // we want to prevent x-pack from adding custom metadata
+            return node.getVersion().before(Version.V_6_3_0) || Booleans.parseBoolean(xpackInstalledAttr) == false;
+        }).collect(Collectors.toList());
+
+        return notReadyNodes;
+    }
+
+    @Override
+    public Settings additionalSettings() {
+        final String xpackInstalledNodeAttrSetting = "node.attr." + XPACK_INSTALLED_NODE_ATTR;
+
+        if (transportClientMode) {
+            if (settings.get(xpackInstalledNodeAttrSetting) != null) {
+                throw new IllegalArgumentException("Directly setting [" + xpackInstalledNodeAttrSetting + "] is not permitted");
+            }
+
+            return super.additionalSettings();
+        } else {
+            if (settings.get(xpackInstalledNodeAttrSetting) != null &&
```

Comment:
why do we allow this setting to be already set in this case? shouldn't we lock it down?

Reply:
I think its because the internal cluster integration test framework will restart nodes with settings copied from the node immediately before it was stopped.  There's a comment [here](https://github.com/elastic/elasticsearch/blob/da6a56f3cc0e343fe51f1ab0faf1da3d026bab91/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java#L319) for the same problem that could be copied over to prevent confusion.

Reply:
exactly. I wanted to fix that behavior, but did not want to make it part of this PR.

Reply:
I see. Makes sense. Thanks.

Reply:
I've added a comment in ba94bdd14af

Reply:
I've opened https://github.com/elastic/elasticsearch/pull/30780 which would allow me to make this check more strict.

Reply:
I've merged that PR and made the check more strict in 1b6ae1d9a59